router/h2/src/main/scala/io/buoyant/router/h2/ClassifiedRetryFilter.scala

```diff
@@ -137,6 +138,11 @@ class ClassifiedRetryFilter(
         case e if responseClassifier(H2ReqRep(req, Throw(e))).contains(ResponseClass.RetryableFailure) =>
           // Request is retryable, attempt to create a new child request stream.
           retry(orElse = { retriesStat.add(count); Future.exception(e) })
+        case NonFatal(e) =>
+          // request is not retryable, we need to make sure the buffer is discarded
+          // so all buffered frames are released
+          requestBuffer.discardBuffer()
```

Comment:
I think we also want to do `retriesStat.add(count)` here, just like in `discardAndReturn`.